crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+    unsigned char calc_pubuf[OSSL_HPKE_MAXSIZE];
+    size_t calc_pubuf_len = OSSL_HPKE_MAXSIZE;
+    point_conversion_form_t form = POINT_CONVERSION_UNCOMPRESSED;
+#endif
+    unsigned char hf_prbuf[OSSL_HPKE_MAXSIZE];
+    size_t hf_prbuf_len = 0;
+
+    if (hpke_kem_id_check(kem_id) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    kem_ind = kem_iana2index(kem_id);
+    if (kem_ind == 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    keytype = hpke_kem_tab[kem_ind].keytype;
+    groupname = hpke_kem_tab[kem_ind].groupname;
+    if (prbuf == NULL || prbuf_len == 0 || retpriv == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (hpke_kem_tab[kem_ind].Npriv == prbuf_len) {
+        if (keytype == NULL) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        param_bld = OSSL_PARAM_BLD_new();
+        if (param_bld == NULL) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        if (groupname != NULL
+            && OSSL_PARAM_BLD_push_utf8_string(param_bld, "group",
+                                               groupname, 0) != 1) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        if (pubuf != NULL && pubuf_len > 0) {
+            if (OSSL_PARAM_BLD_push_octet_string(param_bld, "pub", pubuf,
+                                                 pubuf_len) != 1) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+        } else if (hpke_kem_id_nist_curve(kem_id) == 1) {
+#ifndef OPENSSL_NO_EC
+            /* need to calculate that public value, but we can:-) */
+            groupnid = hpke_kem_tab[kem_ind].groupid;
+            pubsize = hpke_kem_tab[kem_ind].Npk;
+            memset(calc_pubuf, 0, calc_pubuf_len); /* keep asan happy */
+            curve = EC_GROUP_new_by_curve_name(groupnid);
+            if (curve == NULL) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            calc_priv = BN_bin2bn(prbuf, prbuf_len, NULL);
+            if (calc_priv == NULL) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            calc_pub = EC_POINT_new(curve);
+            if (calc_pub == NULL) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            if (EC_POINT_mul(curve, calc_pub, calc_priv, NULL, NULL,
+                             NULL) != 1) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            if ((calc_pubuf_len = EC_POINT_point2oct(curve, calc_pub, form,
+                                                     calc_pubuf, calc_pubuf_len,
+                                                     NULL)) != pubsize) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            if (OSSL_PARAM_BLD_push_octet_string(param_bld, "pub", calc_pubuf,
+                                                 calc_pubuf_len) != 1) {
+                OSSL_HPKE_err;
+                goto err;
+            }
```

Comment:
Yuck. We really should have an EVP API to make this easier. :-(

Reply:
indeed. And failing to do that used lead to a crash. I believe that crash was fixed in https://github.com/openssl/openssl/pull/18902 but still throwing an error if no public key is supplied (says the comment). I'd argue that's unfortunate and it'd be better to recalculate the public key if you can, and you can when given the private key (so that this ickky code to recalculate the public key would only need to be in one place:-)